detectron2_detector/detectron2_detector/detectron2_node.py

```diff
@@ -0,0 +1,181 @@
+
+# import some common detectron2 utilities
+from detectron2 import model_zoo
+from detectron2.engine import DefaultPredictor
+from detectron2.config import get_cfg
+from detectron2.utils.visualizer import Visualizer
+from detectron2.data import MetadataCatalog, DatasetCatalog
+
+import rclpy
+from rclpy.node import Node
+
+from sensor_msgs.msg import Image, PointCloud2 
+from nav2_dynamic_msgs.msg import Obstacle, ObstacleArray
+from geometry_msgs.msg import Pose, Point
+from detectron2_detector.utils import NMS_3D
+
+class Detectron2Detector(Node):
+    '''use Detectron2 to detect object masks from 2D image and estimate 3D position with Pointcloud2 data
+    '''
+    def __init__(self):
+        super().__init__('detectron_node')
+        self.declare_parameters(
+            namespace='',
+            parameters=[
+                ('detectron_config_file', "COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml"),
+                ('detectron_score_thresh', 0.8),
+                ('pointcloud2_topic', "/realsense/camera/pointcloud"),
+                ('pc_downsample_factor', 16),
+                ('min_mask', 20),
+                ('categories', [0]),
+                ('nms_filter', 0.3),
+                ('outlier_filter', 0.5)
+            ])
+        self.pc_downsample_factor = int(self.get_parameter("pc_downsample_factor")._value)
+        self.min_mask = self.get_parameter("min_mask")._value
+        self.categories = self.get_parameter("categories")._value
+        self.nms_filter = self.get_parameter("nms_filter")._value
+        self.outlier_filter = self.get_parameter("outlier_filter")._value
+
+        # setup detectron model
+        self.cfg = get_cfg()
+        config_file = self.get_parameter("detectron_config_file")._value
+        self.cfg.merge_from_file(model_zoo.get_config_file(config_file))
+        self.cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = self.get_parameter("detectron_score_thresh")._value
+        self.cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url(config_file)
+        self.predictor = DefaultPredictor(self.cfg)
+
+        # subscribe to sensor 
+        self.subscription = self.create_subscription(
+            PointCloud2,
+            self.get_parameter("pointcloud2_topic")._value,
+            self.callback,
+            1)
+
+        # setup publisher
+        self.detect_obj_pub = self.create_publisher(ObstacleArray, 'detection', 2)
+        self.detect_img_pub = self.create_publisher(Image, 'image', 2)
+
+        self.count = -1
+
+    def outlier_filter(self, x, y, z, idx):
+        '''simple outlier filter, assume Gaussian distribution and drop points with low probability (too far away from center)'''
+        mean = [np.mean(x), np.mean(y), np.mean(z)]
+        cov = np.diag([np.var(x), np.var(y), np.var(z)])
+        rv = multivariate_normal(mean, cov)
+        points = np.dstack((x, y, z))
+        p = rv.pdf(points)
+        return idx[p > self.outlier_filter]
+
+    def callback(self, msg):
```

Comment:
This function is 1 really long method, can you split it up into a couple of individual functions?

Reply:
Both for readability, and so that we can add tests to some parts of this code that isn't the literal "detect" line